### PR TITLE
Feature/testing v2

### PR DESCRIPTION
**Pull request by drewbanin**
Testing V2 -- This makes it possible to test that model schemas conform to expectations

Here are two example schema definitions for `pardot_visitoractivity` and `pardot_visitoractivity_events_meta`.

File: `models/pardot/schema.yml`
```yml
pardot_visitoractivity:
  constraints:
    not_null:
      - prospect_id
    unique: []
    relationships:
      - {from: type, to: pardot_visitoractivity_events_meta, field: type}
      - {from: type, to: pardot_visitoractivity_types_meta, field: type}

pardot_visitoractivity_events_meta:
  constraints:
    not_null:
      - type_name
      - event_name
    unique:
      - event_name
    relationships:
      - {from: type, to: pardot_visitoractivity, field: type}
```

There are three possible validations:
 - uniqueness 
 - not-null
 - referential integrity

For more information: https://github.com/analyst-collective/dbt/issues/38

### Usage
`dbt test` : run V1 tests (temp create views w/ 0 rows -- checks for sql validity)
`dbt test --validate` : run V2 tests (query for rows which do not conform to the specified schema)
`dbt test --validate --skip-test-creates` : run V2 tests, but skip V1 tests 

Note that testing also follows the enabled/disabled rules in `dbt_config.yml`. For instance, if `pardot` models are disabled, then the pardot tests will not be run!

### CODE REVIEW

**Comment by jthandy** [comment]
@drewbanin is it required to create the empty list if there aren't any constraints of a certain type? can't you just not mention unique at all in that example? i wouldn't want a bunch of random shit cluttering up the config...

**Comment by jthandy** [comment]
@drewbanin how do schema files work? can they be nested at any level? do they have to be in a first-level folder?

File: dbt/main.py [hunk 1]
```diff
@@ -40,6 +40,8 @@ def main(args=None):
     sub.set_defaults(cls=run_task.RunTask, which='run')
 
     sub = subs.add_parser('test', parents=[base_subparser])
+    sub.add_argument('--skip-test-creates', action='store_true', help="Don't create temporary views to validate model SQL")
+    sub.add_argument('--validate', action='store_true', help='Name of the new project')
```
Comment (jthandy) [inline dbt/main.py]: fix help text

File: dbt/schema_tester.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+                    cursor.execute(sql)
+                except Exception as e:
+                    e.model = model
+                    raise e
+
+                result = cursor.fetchone()
+                if len(result) != 1:
+                    print("SQL: {}".format(sql))
+                    print("RESULT:".format(result))
+                    raise RuntimeError("Unexpected validation result. Expected 1 record, got {}".format(len(result)))
```
Comment (jthandy) [inline dbt/schema_tester.py]: expected 0 results, right?

File: dbt/schema_tester.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        return {
+            "schema": schema,
+            "table": table,
+            "field": field
+        }
+
+    def make_query(self, query, params):
+        return query.format(**params)
+
+    def get_target(self):
```
Comment (jthandy) [inline dbt/schema_tester.py]: This seems weird. Why are we doing this in this specific file? Why isn't this being done at some more global level that can be used in testing and execution?

File: dbt/schema_tester.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
+import os
+import yaml
+
+from dbt.runner import RedshiftTarget
+
+QUERY_VALIDATE_NOT_NULL = """
```
Comment (jthandy) [inline dbt/schema_tester.py]: maybe split these out into some config? seems out of place to have them here...

**Comment by jthandy** [comment]
Made some comments. If you think that any of them are too onerous and shouldn't be taken care of, I'm fine with merging this as long as you're confident that you've tested the functionality.

File: dbt/schema_tester.py (hunk 2 above)
Comment (drewbanin) [inline dbt/schema_tester.py]: Naw, in this case we only want one result: The number of records that failed the validation. This is a sanity check for the following type of query:

```sql
with validation as (
  select "{field}" as f
  from "{schema}"."{table}"
)
select count(*) from validation where f is null
```

File: dbt/schema_tester.py (hunk 3 above)
Comment (drewbanin) [inline dbt/schema_tester.py]: Yeah, you're totally right. This function was originally in the Runner class but I duplicated it here. Not great. I can take a look at breaking it out into some sort of Util file that they both use

File: dbt/schema_tester.py (hunk 4 above)
Comment (drewbanin) [inline dbt/schema_tester.py]: there's a `templates.py` file that's probably a good place for this. I'll swizzle it aroudn

**Comment by drewbanin** [comment]
- constraint keys are optional, so you can supply `not_null: []` or just omit the `not_null` key. I can update the example accordingly
- schema files must be in the top-level directory (eg. pardot/) for now, but that will change (so it can be anywhere) when I refactor the existing multi-level-nesting problems

Great feedback, thanks!

**Pull request merged**